File: vm.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>

#define arr(m) (m?(uint*)m:zero)
#define C w & 7
#define B (w >> 3) & 7
#define A (w >> 6) & 7

typedef unsigned int uint;

static uint * ulloc(uint size) {
    uint * r = (uint*)calloc((1 + size), 4);
    *r = size;
    return (r + 1);
}

int main (int argc, char ** argv) {
    static uint reg[8], ip, * zero;
    FILE * f = fopen(argv[1], "rb");
    if (!f) return -1;
    struct stat buf;
    if (stat(argv[1], &buf)) return -1;
    else zero = ulloc(buf.st_size >> 2);
    int a, n = 4, i = 0;
    while(EOF != (a = fgetc(f))) {
        if (!n--) { i++; n = 3; }
        zero[i] = (zero[i] << 8) | a;
    }
    fclose(f);
    for(;;) {
        uint w = zero[ip++];
        switch(w >> 28) {
        case 0: if (reg[C]) reg[A] = reg[B]; break;  /* move */
        case 1: reg[A] = arr(reg[B])[reg[C]]; break; /* array index */
        case 2: arr(reg[A])[reg[B]] = reg[C]; break; /* array amendment */
        case 3: reg[A] = reg[B] + reg[C]; break;     /* addition */
        case 4: reg[A] = reg[B] * reg[C]; break;     /* multiplication */
        case 5: reg[A] = reg[B] / reg[C]; break;     /* division */
        case 6: reg[A] = ~(reg[B] & reg[C]); break;  /* not-and */
        case 7: return 0;                            /* halt */
        case 8: reg[B] = (uint)ulloc(reg[C]); break; /* allocation */
        case 9: free(-1 + (uint*)reg[C]); break;     /* abandonment */
        case 10: putchar(reg[C]); break;             /* output */
        case 11: reg[C] = getchar(); break;          /* input */
        case 12:                                     /* load program */
            if (reg[B]) {
                free(zero - 1);
                int size = ((uint*)reg[B])[-1];
                zero = ulloc(size);
                memcpy(zero, (uint*)reg[B], size * 4);
            }
            ip = reg[C];
            break;
        case 13: reg[7 & (w >> 25)] = w & 0177777777; /* load immediate */
        }
    }
}

File: vm.rb
#!/usr/bin/env ruby

class VM

  attr_accessor :arrays, :regs, :pc

  def initialize(program_file)
    @arrays = []
    @regs = [0, 0, 0, 0, 0, 0, 0]
    @pc = 0
    load_program(program_file)
  end

  def load_program(program_file)
    @arrays[0] = IO.read(program_file).unpack("N*")
  end

  def check_oob(array_reg, index_reg)
    array = @regs[array_reg]
    index = @regs[index_reg]
    if array >= @arrays.size
      $stderr.puts "illegal array number #{array} (there are only #{@arrays.size} arrays)"
      return
    end
    if @arrays[array].nil?
      $stderr.puts "array[#{array}]is nil"
      return
    end
    if index >= @arrays[array].size
      $stderr.puts "illegal array[#{array}] index #{index} (size of array[#{array}] is #{@arrays[array].size})"
    end
  end

  def run
    while true
      $stderr.puts to_debug_s() if $DEBUG
      ins = @arrays[0][@pc]
      @pc += 1
      a, b, c = ((ins >> 6) & 7), ((ins >> 3) & 7), (ins & 7)
      case (ins >> 28)
      when 0                    # move
        @regs[a] = @regs[b] if @regs[c] != 0
      when 1                    # array index
        check_oob(b, c) if $DEBUG
        @regs[a] = @arrays[@regs[b]][@regs[c]]
      when 2                    # array amendment
        check_oob(a, b) if $DEBUG
        @arrays[@regs[a]][@regs[b]] = @regs[c]
      when 3                    # addition
        @regs[a] = (@regs[b] + @regs[c]) & 0xffffffff
      when 4                    # multiplication
        @regs[a] = (@regs[b] * @regs[c]) & 0xffffffff
      when 5                    # division
        @regs[a] = (@regs[b] / @regs[c]) & 0xffffffff
      when 6                    # not-and
        @regs[a] = ~(@regs[b] & @regs[c]) & 0xffffffff
      when 7                    # halt
        exit 0
      when 8                    # allocation
        # Weird: if I use @regs[c] in the Array.new call below, some values
        # (those above 32768?) get changed to a lower value like 2. Reading
        # reg C here fixes that problem.
        size = @regs[c]
        @regs[b] = @arrays.index(nil) || @arrays.size
        @arrays[@regs[b]] = Array.new(size, 0)
      when 9                    # abandonment
        @arrays[@regs[c]] = nil
      when 10                   # output
        print @regs[c].chr
        # $stdout.flush
      when 11                   # input
        @regs[c] = getch
      when 12                   # load program
        if @regs[b] > 0
          @arrays[0] = @arrays[@regs[b]].dup
        end
        @pc = @regs[c]
      when 13                   # load immediate
        @regs[(ins >> 25) & 7] = ins & 0x1ffffff
        $stderr.puts "\tloadi r[#{(ins >> 25) & 7}] = #{ins & 0x1ffffff}" if $DEBUG
      else
        $stderr.puts "illegal instruction"
        exit 1
      end
    end
  end

  def to_debug_s(decr_pc=false)
    @pc -= 1 if decr_pc
    ins = @arrays[0][@pc]
    a, b, c = ((ins >> 6) & 7), ((ins >> 3) & 7), (ins & 7)
    opcode = (ins >> 28)
    str = "pc = #{'%5d' % pc}"
    str << ", ins = 0x#{'%08x' % ins} 0#{'%011o' % ins}"
    str << ", op = #{'%2d' % opcode}"
    str << ", a = #{a}, b = #{b}, c = #{c}"
    str << ", ra = #{@regs[a]}, rb = #{@regs[b]}, rc = #{@regs[c]}"
    str << " regs = #{regs}"
    str << ' "'
    @regs.each { |val| str << (val.nil? ? "*" : (val >= 32 && val <= 126 ? val.chr : '.')) }
    str << '"'
    @pc += 1 if decr_pc
    str
  end

end

if __FILE__ == $0 && ARGV[0]
  VM.new(ARGV[0]).run
end
